src/libraries/System.Private.CoreLib/src/System/Text/CompositeFormat.cs

```diff
@@ -0,0 +1,353 @@
+            }
+        }
+
+        /// <summary>Parse the composite format string into segments.</summary>
+        /// <param name="format">The format string.</param>
+        /// <param name="segments">The list into which to store the segments.</param>
+        /// <returns>true if the format string can be parsed successfully; otherwise, false.</returns>
+        private static bool TryParseLiterals(ReadOnlySpan<char> format, List<(string? Literal, int ArgIndex, int Alignment, string? Format)> segments)
+        {
+            // This parsing logic is copied from string.Format.  It's the same code modified to not format
+            // as part of parsing and instead store the parsed literals and argument specifiers (alignment
+            // and format) for later use.
+
+            // Rather than parsing directly into the segments list, literals are parsed into a reusable builder.
+            // Due to the nature of the parsing logic copied from string.Format, and our desire not to veer from
+            // it significantly in order to maintain compatibility and accidental regression, multiple literals
+            // next to each other might be parsed separately due to braces in between them.  This builder then
+            // allows us to merge those segments back together easily prior to their being appended to the list.
+            var vsb = new ValueStringBuilder(stackalloc char[StackAllocCharSize]);
+
+            // Repeatedly find the next hole and process it.
+            int pos = 0;
+            char ch;
+            while (true)
+            {
+                // Skip until either the end of the input or the first unescaped opening brace, whichever comes first.
+                // Along the way we need to also unescape escaped closing braces.
+                while (true)
+                {
+                    // Find the next brace.  If there isn't one, the remainder of the input is text to be appended, and we're done.
+                    ReadOnlySpan<char> remainder = format.Slice(pos);
+                    int countUntilNextBrace = remainder.IndexOfAny('{', '}');
+                    if (countUntilNextBrace < 0)
+                    {
+                        vsb.Append(remainder);
+                        segments.Add((vsb.ToString(), -1, 0, null));
+                        return true;
+                    }
+
+                    // Append the text until the brace.
+                    vsb.Append(remainder.Slice(0, countUntilNextBrace));
+                    pos += countUntilNextBrace;
+
+                    // Get the brace.  It must be followed by another character, either a copy of itself in the case of being
+                    // escaped, or an arbitrary character that's part of the hole in the case of an opening brace.
+                    char brace = format[pos];
+                    if (!TryMoveNext(format, ref pos, out ch))
+                    {
+                        return false;
```

Comment:
>return false; [](http://example.com/codeflow?start=24&length=13)

We cannot have a single `}` at the end of the format?

Reply:
reading more the rest of the code, I think I got the idea. we shouldn't encounter `}` without escaping in this loop.